src/vs/editor/common/model/treeSitterTokens.ts

```diff
@@ -0,0 +1,226 @@
+		super();
+		this._register(this._textModel.onDidChangeContent(() => this._onDidChangeContent()));
+	}
+
+	private _onDidChangeContent() {
+		// TODO @alexr00 this is an extremely naive implementation and should absolutely not be pushed to main, it's just a placeholder.
+		this._tree?.delete();
+		this._tree = undefined;
+	}
+
+	private _ensureTree() {
+		if (!this._tree) {
+			this._tree = this._parser.parse(this._textModel.getValue());
+		}
+		return this._tree;
+	}
+
+	private _ensureQuery() {
+		if (!this._query) {
+			this._query = this._language.query(this._queries);
+		}
+		return this._query;
+	}
+
+	/**
+	 * Gets the tokens for a given line.
+	 * Each token takes 2 elements in the array. The first element is the offset of the end of the token *in the line, not in the document*, and the second element is the metadata.
+	 *
+	 * @param lineNumber
+	 * @returns
+	 */
+	public lineTokens(lineNumber: number): Uint32Array {
+		const tree = this._ensureTree();
+		const query = this._ensureQuery();
+		const lineLength = this._textModel.getLineMaxColumn(lineNumber);
+		const captures = query.captures(tree.rootNode, { startPosition: { row: lineNumber - 1, column: 0 }, endPosition: { row: lineNumber - 1, column: lineLength - 1 } });
+		if (captures.length === 0 && lineLength > 0) {
+			// No captures, but we always want to return at least one token for each line
+			return new Uint32Array([lineLength, 0]);
+		}
+
+		let tokens: Uint32Array = new Uint32Array(captures.length * 2);
+		let tokenIndex = 0;
+		for (let captureIndex = 0; captureIndex < captures.length; captureIndex++) {
+			const capture = captures[captureIndex];
+			const metadata = this.findMetadata(capture.name);
+			const lineStartOffset = this._textModel.getOffsetAt({ lineNumber: lineNumber, column: 0 });
+			const offset = this._textModel.getOffsetAt({ lineNumber: capture.node.endPosition.row + 1, column: capture.node.endPosition.column + 1 }) - lineStartOffset;
+			// Not every character will get captured, so we need to make sure that our current capture doesn't bleed toward the start of the line and cover characters that it doesn't apply to.
+			// We do this by creating a new token in the array if the previous token ends before the current token starts.
+			if (captureIndex > 0) {
+				const previousTokenEnd = tokens[(tokenIndex - 1) * 2];
+				const currentTokenLength = capture.node.text.length;
+				const intermediateTokenOffset = offset - currentTokenLength;
+				if (previousTokenEnd < intermediateTokenOffset) {
+					tokens[tokenIndex * 2] = intermediateTokenOffset;
+					tokens[tokenIndex * 2 + 1] = 0;
+					tokenIndex++;
+					const newTokens = new Uint32Array(tokens.length + 2);
+					newTokens.set(tokens);
+					tokens = newTokens;
+				}
+			}
+			tokens[tokenIndex * 2] = offset;
+			tokens[tokenIndex * 2 + 1] = metadata;
+			tokenIndex++;
+		}
+		return tokens;
+	}
+
+	private findMetadata(captureName: string): number {
```

Comment:
@hediet, do you know if there's a better way to retrieve the token metadata given a scope? Relying on the `ColorThemeData` is layer breaking. I started trying to move things around to fix the layer breaker, but it quickly got complicated.